homestar-wasm/README.md

```diff
@@ -23,15 +23,1055 @@
+    and passed into `fn` as a `list<s32>` argument.
+
+  * **WIT to IPLD Translation**:
+
+    Conversely, when a `list` value is returned from a WIT function, it is
+    translated back into an `Ipld::List` value.
+
+- [`tuple`][wit-tuple]:
+
+  * **IPLD to WIT Translation**
+
+    When a WIT function expects a `tuple` input, an `Ipld::List` value is
+    mapped to a `tuple` WIT runtime value.
+
+    **Example**:
+
+    ```wit
+    type ipv6-socket-address = tuple<u16, u16, u16, u16, u16, u16, u16, u16>;
+
+    export fn: func(a: ipv6-socket-address) -> tuple<u32, u32>;
+    ```
+
+    Given an JSON input for this function:
+
+    ```json
+    {
+      "args": [[8193, 3512, 34211, 0, 0, 35374, 880, 29492]]
+    }
+    ```
+
+    `[8193, 3512, 34211, 0, 0, 35374, 880, 29492]` is converted into an
+    `Ipld::List`, which is then translated and passed into `fn` as a
+    `tuple<u16, u16, u16, u16, u16, u16, u16, u16>` argument.
+
+    **If the length of list does not match not match the number of fields in the
+    tuple interface type, then an error will be thrown in the interpreter.**
+
+  * **WIT to IPLD Translation**:
+
+    Conversely, when a `tuple` value is returned from a WIT function, it is
+    translated back into an `Ipld::List` value.
+
+- [`flags`][wit-flags]:
+
+  `flags` represent a bitset structure with a name for each bit. The type
+  represents a set of named booleans. In an instance of the type, each flag will
+  be either true or false.
+
+  * **IPLD to WIT Translation**
+
+    When a WIT function expects a `flags` input, an `Ipld::List` value is
+    mapped to a `flags` WIT runtime value.
+
+    When used as an input, you can set the flags you want turned on/true as an
+    inclusive subset of strings. When used as an output, you will get a list of
+    strings representing the flags that are set to true.
+
+
+    **Example**:
+
+    ```wit
+    flags permissions {
+        read,
+        write,
+        exec,
+    }
+
+    export fn: func(perm: permissions) -> bool;
+    ```
+
+    Given an JSON input for this function:
+
+    ```json
+    {
+      "args": [["read", "write"]]
+    }
+    ```
+
+    `[read, write]` is converted into an `Ipld::List`, which is then translated
+    and passed into `fn` as a `permissions` argument.
+
+  * **WIT to IPLD Translation**:
+
+    Conversely, when a `flags` value is returned from a WIT function, it is
+    translated back into an `Ipld::List` value.
+
+
+**IPLD Schema Definitions**:
+
+``` ipldsch
+type IPLDListAsWit union {
+  | List [any]
+  | Tuple [any]
+  | Flags [string]
+} representation kinded
+
+type WitAsIpldList union {
+  | List [any]
+  | Tuple [any]
+  | Flags [string]
+} representation kinded
+```
+
+#### Maps
+
+This section outlines the translation process between IPLD map values
+(`Ipld::Map`) and varying [WIT runtime values][wit-val]. A `Ipld::Map`
+value can be interpreted as one of a `record`, `variant`, or
+a `list` of 2-element `tuples`.
```

Comment:
```suggestion
a `list` of two-element `tuples`.
```
Somewhat of a stylistic choice, but numbers less that ten are usually written out.